pandas/io/formats/csvs.py

```diff
@@ -33,27 +35,26 @@ class CSVFormatter:
     def __init__(
         self,
         obj,
-        path_or_buf=None,
-        sep=",",
-        na_rep="",
+        path_or_buf: Optional[FilePathOrBuffer[str]] = None,
+        sep: str = ",",
+        na_rep: str = "",
         float_format=None,
         cols=None,
-        header=True,
-        index=True,
+        header: Union[bool, Sequence[str]] = True,
+        index: bool = True,
         index_label=None,
-        mode="w",
-        encoding=None,
+        mode: str = "w",
+        encoding: Optional[str] = None,
         compression="infer",
```

Comment:
```suggestion
        compression: str = "infer",
```

Reply:
Should it be typed as
```python
Union[str, Mapping[str, str], None]
```

?

---

mypy's log for 2 diffrent annotations:

```python
compression: str = "infer"
```

```
$ mypy pandas
pandas/io/formats/csvs.py:63: error: Incompatible types in assignment (expression has type "Optional[str]", variable has type "str")
pandas/core/generic.py:3167: error: Argument "compression" to "CSVFormatter" has incompatible type "Union[str, Mapping[str, str], None]"; expected "str"
Found 2 errors in 2 files (checked 816 source files)
```

---

```python
compression: Union[str, Mapping[str, str], None] = "infer"
```

```
$ mypy pandas
Success: no issues found in 816 source files
```

Reply:
Hmm OK I guess should match as `compression: Optional[Union[str, Mapping[str, str]]] = "infer",`